javaslang/src/main/java/javaslang/collection/Vector.java

```diff
@@ -783,7 +824,51 @@ public boolean isTraversableAgain() {
 
     @Override
     public Iterator<T> iterator() {
-        return isEmpty() ? Iterator.empty() : trie.iterator();
+        return isEmpty() ? Iterator.empty()
+                         : new VectorIterator();
+    }
+
```

Comment:
My main pain point in previous discussions was here that we leak BMT internals into the Vector impl.

How about leaving the BMT.iterator() method as is and implementing Vector.iterator() like that:

```java
@Override
public Iterator<T> iterator() {
    return (Iterator<T>) Iterator.concat(
            Iterator.of(leading),
            trie.iterator(),
            Iterator.of(trailing)
    );
}
```

_(Did not test it)_